rfqservice/order.go

```diff
@@ -0,0 +1,313 @@
+	if htlc.AmountOutMsat <= c.MinimumChannelPayment {
+		return fmt.Errorf("htlc out amount is less than the remit's "+
+			"minimum (htlc_out_msat=%d, remit_min_msat=%d)",
+			htlc.AmountOutMsat, c.MinimumChannelPayment)
+	}
+
+	// Lastly, check to ensure that the channel remit has not expired.
+	if time.Now().Unix() > int64(c.Expiry) {
+		return fmt.Errorf("channel remit has expired (expiry=%d)",
+			c.Expiry)
+	}
+
+	return nil
+}
+
+// OrderHandlerCfg is a struct that holds the configuration parameters for the
+// order handler service.
+type OrderHandlerCfg struct {
+	// CleanupInterval is the interval at which the order handler cleans up
+	// expired accepted quotes from its local cache.
+	CleanupInterval time.Duration
+
+	// HtlcInterceptor is the HTLC interceptor. This component is used to
+	// intercept and accept/reject HTLCs.
+	HtlcInterceptor HtlcInterceptor
+}
+
+// OrderHandler orchestrates management of accepted quote bundles. It monitors
+// HTLCs (Hash Time Locked Contracts), and determines acceptance/rejection based
+// on the terms of the associated accepted quote.
+type OrderHandler struct {
+	startOnce sync.Once
+	stopOnce  sync.Once
+
+	// cfg holds the configuration parameters for the RFQ order handler.
+	cfg OrderHandlerCfg
+
+	// channelRemits is a map of serialised short channel IDs (SCIDs) to
+	// associated active channel remits.
+	channelRemits map[SerialisedScid]ChannelRemit
+
+	// channelRemitsMtx guards the channelRemits map.
+	channelRemitsMtx sync.Mutex
+
+	// ContextGuard provides a wait group and main quit channel that can be
+	// used to create guarded contexts.
+	*fn.ContextGuard
+}
+
+// NewOrderHandler creates a new struct instance.
+func NewOrderHandler(cfg OrderHandlerCfg) (*OrderHandler, error) {
+	return &OrderHandler{
+		cfg:           cfg,
+		channelRemits: make(map[SerialisedScid]ChannelRemit),
+		ContextGuard: &fn.ContextGuard{
+			DefaultTimeout: DefaultTimeout,
+			Quit:           make(chan struct{}),
+		},
+	}, nil
+}
+
+// handleIncomingHtlc handles an incoming HTLC.
+//
+// NOTE: This function must be thread safe.
+func (h *OrderHandler) handleIncomingHtlc(_ context.Context,
+	htlc lndclient.InterceptedHtlc) (*lndclient.InterceptedHtlcResponse,
+	error) {
+
+	log.Debug("Handling incoming HTLC")
+
+	// TODO(ffranr): I think we need a way to silently dismiss HTLCs that
+	//  are not associated with tap.
+
+	scid := SerialisedScid(htlc.OutgoingChannelID.ToUint64())
+	channelRemit, ok := h.FetchChannelRemit(scid)
+
+	// If a channel remit does not exist for the channel SCID, we reject the
+	// HTLC.
+	if !ok {
+		return &lndclient.InterceptedHtlcResponse{
+			Action: lndclient.InterceptorActionFail,
+		}, nil
+	}
+
+	// At this point, we know that the channel remit exists and has not
+	// expired whilst sitting in the local cache. We can now check that the
+	// HTLC complies with the channel remit.
+	err := channelRemit.CheckHtlcCompliance(htlc)
```

Comment:
Also note that we'll need to distinguish between crossing _into_ BTC and _out_ of it. When we cross in, we're receiving an asset and sending out BTC, when we cross out the opposite. 

The current structs and flows seem to be concerned mostly with the out (recv flow for leaf node), but we also need to be concerned with the opposite. I think _this_ is where the distinction between an ask and a bid exists:
 * When we receive in an asset to send out BTC, we're effectively buying the asset and selling BTC. 
 * When we receive in BTC to send out an asset, we're effectively selling the asset and buying BTC.